climakitae/util/utils.py

```diff
@@ -448,3 +451,68 @@ def hdh_cdh_lineplot(data):
     return data.hvplot.line(
         x="time", by="simulation", title=data.name, ylabel=data.name + " (degF)"
     )
+
+
+def convert_to_local_time(
+    data: xr.Dataset, selections: Select, lat: float, lon: float
+) -> xr.Dataset:
+    """
+    Converts the inputted data to the local time of the selection.
+    """
+    # 1. Find the other data
+    start, end = selections.time_slice
+    tz_selections = copy.copy(selections)
+    tz_selections.time_slice = (
+        end + 1,
+        end + 1,
+    )  # This is assuming selections passed with be negative UTC time. Also to get the next year of data.
+
+    print("Retrieving data...")
+    tz_data = tz_selections.retrieve()
+
+    # 2. Combine the data
+    total_data = xr.concat([data, tz_data], dim="time")
+
+    # 3. Change datetime objects to local time
+    print("Converting to local time...")
+    tf = TimezoneFinder()
+    local_tz = tf.timezone_at(lat=lat, lng=lon)  # Finding local time zone
+    dt_index = xr.CFTimeIndex(
+        total_data.time.values
+    ).to_datetimeindex()  # Converting the datetime objects into a format that allows for timezone changes (pandas Datetime Index)
+    localized = dt_index.tz_localize("UTC").tz_convert(
+        local_tz
+    )  # Set timestamps to UTC and convert to local time
+
+    # Convert pandas Datetime objects back to cftime objects
+    def convert_to_cftime(pd_dt):
+        # Manually removing leap days
+        if pd_dt.month == 2 and pd_dt.day == 29:
+            return cftime.DatetimeNoLeap(
+                year=pd_dt.year,
+                month=pd_dt.month,
+                day=28,
+                hour=pd_dt.hour,
+                minute=pd_dt.minute,
+                second=pd_dt.second,
+            )
+        return cftime.DatetimeNoLeap(
+            year=pd_dt.year,
+            month=pd_dt.month,
+            day=pd_dt.day,
+            hour=pd_dt.hour,
+            minute=pd_dt.minute,
+            second=pd_dt.second,
+        )
+
+    # Apply function to Datetime objects and assign back to DataArray
+    total_data["time"] = np.array(localized.map(convert_to_cftime))
```

Comment:
Looking at this won't the `datetime64` data be converted to `cftime` for the gridded data? or does the `np.array` convert it back and then the station data will be not `cftime`?

Reply:
Yes this is true. I will need to debug what happens if you select gridded data.

Reply:
I believe this is resolved with the consolidation of only using `datetime64[ns]` time now.